README.md
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)
[![Join the chat at https://gitter.im/DylanPiercey/auto-sni](https://badges.gitter.im/DylanPiercey/auto-sni.svg)](https://gitter.im/DylanPiercey/auto-sni?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

# Auto SNI

SSL Certificates using SNI with almost zero configuration for free with https://letsencrypt.org!

If you have any questions, throw them up on gitter.

# Installation

#### Npm
```console
npm install auto-sni
```

# Features
+ Fetch SSL certificates from letsencrypt.
+ Automatically renew certificates.
+ Forward all incoming http requests to https.

# Example

```javascript
var createServer = require("auto-sni");

var server = createServer({
  email: ..., // Emailed when certificates expire.
  agreeTos: true, // Required for letsencrypt.
  debug: true, // Add console messages and uses staging LetsEncrypt server. (Disable in production)
  domains: ["mysite.com", ["test.com", "www.test.com"]], // List of accepted domain names. (You can use nested arrays to register bundles with LE).
  dir: "~/letsencrypt/etc", // Directory for storing certificates. Defaults to "~/letsencrypt/etc" if not present.
  ports: {
    http: 80, // Optionally override the default http port.
    https: 443 // // Optionally override the default https port.
  }
});

// Server is a "https.createServer" instance.
server.once("listening", ()=> {
  console.log("We are ready to go.");
});

```

### Usage with express.
```js
var createServer = require("auto-sni");
var express      = require("express");
var app          = express();

app.get("/test", ...);

createServer({ email: ..., domains: ..., agreeTos: true }, app);
```

### Usage with koa.
```js
var createServer = require("auto-sni");
var koa          = require("koa");
var app          = koa();

app.use(...);

createServer({ email: ..., domains: ..., agreeTos: true }, app.callback());
```

### Usage with rill.
```js
var createServer = require("auto-sni");
var rill         = require("rill");
var app          = rill();

app.get("/test", ...);

createServer({ email: ..., domains: ..., agreeTos: true }, app.handler());
```

### Usage with hapi.
```js
// Untested (Let me know in gitter if this doesn't work.)
var createServer = require("auto-sni");
var hapi         = require("hapi");
var server       = new hapi.Server();
var secureServer = createServer({ email: ..., domains: ..., agreeTos: true });

server.connection({ listener: secureServer, autoListen: false, tls: true });
```

### Usage with restify.
```js
// Untested (Let me know in gitter if this doesn't work.)
var createServer = require("auto-sni");
var restify      = require("restify");
var app          = restify.createServer({ name: 'myapp', version: '1.0.0' });

app.get("/test", ...);

createServer({ email: ..., domains: ..., agreeTos: true }, app.server);
```

# Dynamic Domains
You can also specify an async function to approve domains like so:

```js
createServer({
  ...,
  domains: (options, cert, cb) => {
    setTimeout(() => cb({ options, cert }), 1000)
  }
})
```

# Root Access
AutoSNI requires access to low level ports 80 (http) and 443 (https) to operate by default.
These ports are typically restricted by the operating system.

In production (on linux servers) you can use the following command to give Node access to these ports.

```console
sudo setcap cap_net_bind_service=+ep `readlink -f \`which node\``
```

For development it's best to set the "ports" option manually to something like:
```js
{
  ports: {
    http: 3001,
    https: 3002
  }
}

// Access server on localhost:3002
```

# Rate Limits
Currently LetsEncrypt imposes some rate limits on certificate creation.
[Click here for the current rate limits.](https://community.letsencrypt.org/t/rate-limits-for-lets-encrypt/6769)

### Contributions

Please use `npm test` for tests and feel free to create a PR!


CHANGES.md
2.3.2 2017-10-05
==================
  * Add `dir` option to readme example.

2.3.0, 2.3.1 2017-07-22
==================
  * Add ability to approve domains with a function.

2.2.1 2017-06-11
==================
  * Update readme regarding root access to privileged ports.

2.2.0 2017-06-05
==================
  * Flatten domains option for compatibility with @1.x

2.1.1 2017-04-13
==================
  * Internal implementation switched to use greenlock-express.
  * No longer falls back to self signed when unable to communicate with letsencrypt.
  * Removed ability to disable https redirection (should just use vanilla http server).
  * Simplified api.

1.5.2 2016-09-21
==================
  * Fix travis ci.

1.5.1 2016-08-26
==================
  * Update letsencrypt rate limit link in docs.
  * Update bluebird.

1.5.0 2016-08-26
==================
  * Fix mkdirp issue and lock down `letsencrypt@1.5.1` until `auto-sni@2.0`.
  * Fixes issue when using `letsencrypt@1.5.8`.

1.3.0 2016-08-01
==================
  * Add ability to change redirect code when `forceSSL` is enabled.

1.2.1 2016-05-16
==================
  * Fix issue with self signed fallback certificate.

1.1.3 2016-05-16
==================
  * Update dependencies (fixes issue with node 6).

1.1.1 2016-04-23 (Will reset renew certificates)
==================
  * Clears certificate when switching between debug modes.
  * Clears certificate when bundle changes (previously bundles were tied to the first domain).
  * Attempt to fix some issues with restify.

1.0.0 / 2016-04-17
==================
  * Remove bundle option.
  * No longer fuzzy match domains with `path-to-regexp`.
  * Allow for nested arrays in `domains` for bundled certificates.


test/main.js
var test = require('tape')
var ngrok = require('ngrok')
var hostile = require('hostile')
var request = require('supertest')
var createServer = require('../')
var TEST_CONFIG = {
  debug: true,
  email: 'autosni.github@gmail.com',
  agreeTos: true,
  // Explicitly use new ports.
  ports: { http: undefined, https: undefined }
}

test('required fields', function (t) {
  t.plan(4)
  t.throws(createServer.bind(null), TypeError, 'options are required')
  t.throws(createServer.bind(null, { agreeTos: true }), /Email is required/, 'email is required')
  t.throws(createServer.bind(null, { email: 'a@b.com' }), /Must agree to LE TOS/, 'tos is required')

  createServer({
    email: 'a@b.com',
    agreeTos: true,
    // We only use ports here to avoid EACCES.
    ports: { http: 3000, https: 3001 }
  })
    .once('error', t.fail)
    .once('listening', function () {
      this.close()
      t.pass('server should start')
    })
})

test('register certificate fallback to unsigned', function (t) {
  t.plan(4)

  // Handler argument.
  createServer(TEST_CONFIG, helloWorld)
    .once('error', t.fail)
    .once('listening', function () {
      request(this)
        .get('/')
        .set('Host', 'test.com')
        .end(function (err, res) {
          t.ok(err, 'should have an error')
          t.equal(err.message, 'self signed certificate', 'should be a self signed certificate')
          this.close()
        }.bind(this))
    })

  // Request event.
  createServer(TEST_CONFIG)
    .on('request', helloWorld)
    .once('error', t.fail)
    .once('listening', function () {
      request(this)
        .get('/')
        .set('Host', 'test.com')
        .end(function (err, res) {
          t.ok(err, 'should have an error')
          t.equal(err.message, 'self signed certificate', 'should be a self signed certificate')
          this.close()
        }.bind(this))
    })
})

/** These tests are a WIP currently ngrok requires Pro to use custom TLS so an alternative is needed */
test('register certificate with letsencrypt', function (t) {
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
  t.plan(1)
  createServer(TEST_CONFIG, helloWorld)
    .once('error', t.fail)
    .once('listening', function () {
      var server = this
      ngrok.connect({ port: server.http.address().port }, function (err, url) {
        if (err) return t.fail(err)
        var host = url.replace('https://', '')

        hostile.set('127.0.0.1', host, function (err) {
          if (err) return t.fail(err)
          request(url + ':' + server.address().port)
            .get('/')
            .end(function (err, res) {
              if (err) return t.fail(err)
              t.equal(res.text, 'Hello World\n', 'should respond to https request')
              ngrok.disconnect(url)
              server.close()
              hostile.remove('127.0.0.1', host)
            })
        })
      })
    })
})

/**
 * Generic hello world server responder.
 */
function helloWorld (req, res) {
  res.writeHead(200, {'Content-Type': 'text/plain'})
  res.end('Hello World\n')
}


test/index.js
var path = require('path')
var tape = require('tape')

tape
  .createStream()
  .pipe(require('tap-spec')())
  .pipe(process.stdout)

tape.onFinish(process.exit)

process.argv.slice(2).forEach(function (file) {
  require(path.resolve(__dirname, file))
})


test/main.test.js
var test = require('tape')
var ngrok = require('ngrok')
var hostile = require('hostile')
var request = require('supertest')
var assign = require('object-assign')
var createServer = require('../')
test.onFinish(process.exit)

var TEST_CONFIG = {
  debug: true,
  email: 'autosni.github@gmail.com',
  agreeTos: true,
  // Explicitly use new ports.
  ports: { http: null, https: null }
}

test('required fields', function (t) {
  t.plan(6)
  t.throws(createServer.bind(null), TypeError, 'options are required')
  t.throws(createServer.bind(null, { agreeTos: true }), /Email is required/, 'email is required')
  t.throws(createServer.bind(null, { email: 'a@b.com' }), /Must agree to LE TOS/, 'tos is required')
  t.throws(createServer.bind(null, { agreeTos: true, email: 'a@b.com' }), /Domains option must be an array or a function/, 'domains is required')
  t.throws(createServer.bind(null, { agreeTos: true, email: 'a@b.com', domains: [] }), /You must specify at least one domain/, 'domains is required')

  createServer({
    email: 'a@b.com',
    agreeTos: true,
    domains: ['localhost'],
    // We only use ports here to avoid EACCES.
    ports: { http: 3000, https: 3001 }
  })
    .once('error', t.fail)
    .once('listening', function () {
      this.close()
      t.pass('server should start')
    })
})

/** These tests are a WIP currently ngrok requires Pro to use custom TLS so an alternative is needed */
test('register certificate with letsencrypt', function (t) {
  t.plan(1)
  var testPorts = { http: 3003, https: 3004 }
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'

  // Get a free subdomain.
  ngrok.connect(testPorts.http, function (err, url) {
    console.log(err)
    if (err) return t.fail(err)
    var host = url.replace('https://', '')

    // Set alias for host through local machine.
    hostile.set('127.0.0.1', host, function (err) {
      if (err) return t.fail(err)

      // Setup server
      createServer(assign({}, TEST_CONFIG, { ports: testPorts, domains: [host] }), helloWorld)
        .once('error', t.fail)
        .once('listening', function () {
          var server = this
          // Make a request to the server.
          request(url + ':' + server.address().port)
            .get('/')
            .end(function (err, res) {
              if (err) return t.fail(err)
              t.equal(res.text, 'Hello World\n', 'should respond to https request')
              ngrok.kill()
              hostile.remove('127.0.0.1', host)
              server.close()
            })
        })
    })
  })
})

/**
 * Generic hello world server responder.
 */
function helloWorld (req, res) {
  res.writeHead(200, {'Content-Type': 'text/plain'})
  res.end('Hello World\n')
}


lib/load.js
'use strict'

var fs = require('mz/fs')
var cp = require('mz/child_process')
var Promise = require('bluebird')
var mkdirp = require('mkdirp-then')
var path = require('path')
var safe = require('safetimeout')
var LE = require('letsencrypt')
var CONSTANTS = require('./constants')
var STATIC_PATH = CONSTANTS.STATIC_PATH
var CHALLENGE_PATH = CONSTANTS.CHALLENGE_PATH
var EXPIRE_TIME = CONSTANTS.EXPIRE_TIME
var lock = {}
module.exports = load

// Handles for ACME challenges.
var handlers = {
  setChallenge: function (args, key, value, done) {
    var keyfile = path.join(CHALLENGE_PATH, key)
    return fs.writeFile(keyfile, value, 'utf8')
      .then(done.bind(null))
      .catch(done)
  },
  removeChallenge: function (args, key, done) {
    var keyfile = path.join(CHALLENGE_PATH, key)
    return fs.unlink(keyfile)
      .then(done.bind(null))
      .catch(done)
  }
}

/**
 * Attempts to register a hostname with letsencrypt.
 *
 * @param {Object} opts
 * @return {Promise}
 */
function load (domains, opts) {
  var hostname = domains[0]
  // Lock to ensure we are not fetching a host multiple times.
  if (lock[hostname]) return lock[hostname]
  var stdio = opts.debug ? 'pipe' : 'ignore'
  var configDir = path.join(process.env.HOME ? process.env.HOME : process.env.USERPROFILE, '/letsencrypt/etc')
  var liveDir = path.join(configDir, 'live', hostname)
  var liveOptsDir = path.join(liveDir, 'opts.json')
  var logsDir = path.join(configDir, 'letsencrypt.logs')
  var workDir = path.join(configDir, 'letsencrypt.work')
  var args = {
    manual: true,
    debug: opts.debug,
    server: opts.debug ? LE.stagingServerUrl : LE.productionServerUrl,
    configDir: configDir,
    privkeyPath: defaults('privkeyPath'),
    chainPath: defaults('chainPath'),
    fullchainPath: defaults('fullchainPath'),
    certPath: defaults('certPath')
  }

  var hostLock = Promise.all([
    // Load certs.
    fs.readFile(args.privkeyPath, 'ascii'),
    fs.readFile(args.fullchainPath, 'ascii'),
    // Check cert options (debug/bundle)
    fs.readFile(liveOptsDir, 'utf8'),
    // Check cert expiry.
    fs.stat(args.fullchainPath)
  ])
    .then(function checkCertExpiryAndOptions (data) {
      var key = data[0]
      var cert = data[1]
      var prevOpts = JSON.parse(data[2])
      var stat = data[3]
      var ttl = EXPIRE_TIME - (+new Date() - stat.mtime)

      // Check for expired cert.
      if (ttl <= 0) throw new Error('AutoSNI: Certificate Expired.')
      // Check if we are switching from debug to live or vice versa.
      if (prevOpts.debug !== opts.debug) throw new Error('AutoSNI: Switching debug modes.')
      // Check if the domains we have is different.
      if (prevOpts.domains !== String(domains)) throw new Error('AutoSNI: Domain bundle changed.')

      // Remove cert later.
      safe.setTimeout(function () { lock[hostname] = null }, ttl)
      // Return an existing cert.
      return { key: key, cert: cert }
    })
    .catch(function registerOrRenewCert () {
      return mkdirp(liveDir)
        .then(function createFallBackCert () {
          var curOpts = JSON.stringify({ domains: String(domains), debug: opts.debug })
          // Create a new self signed cert, privkey and chain.
          // Also stores some important options that can require recreating a cert.
          cp.execSync(
            // Create the private key and a self signed cert.
            'openssl req -x509 -newkey rsa:2048 -keyout privkey.pem -out cert.pem -days 90 -nodes -subj "' +
              '/CN=' + hostname +
              '/emailAddress=' + opts.email +
              '"',
            { cwd: liveDir, stdio: stdio }
          );
          var privKey = fs.readFileSync(path.join(liveDir, 'privkey.pem'));
          var cert = fs.readFileSync(path.join(liveDir, 'cert.pem'));
          fs.writeFileSync(path.join(liveDir, 'chain.pem'), privKey + cert);
          fs.writeFileSync(path.join(liveDir, 'fullchain.pem'), privKey + cert);
          fs.writeFileSync(path.join(liveDir, 'opts.json'), curOpts);
          return 
        })
        .then(function createLetsEncryptCert () {
          return new Promise(function (resolve, reject) {
            LE.create(args, handlers).register({
              agreeTos: true,
              duplicate: true, // This is only called if the cert is missing anyway.
              domains: domains,
              email: opts.email,
              webrootPath: STATIC_PATH,
              http01Port: opts.ports.http,
              tlsSni01Port: opts.ports.https,
              logsDir: logsDir,
              workDir: workDir
            }, function (err, info) {
              if (err) reject(err)
              else resolve(info)
            })
          })
        })
    })
    .catch(function logError (err) {
      console.error('AutoSNI LetsEncrypt Error:', err)
      console.error('Will fall back to self signed certificate.')
    })
    .then(function maybeRetryLoadCert (info) {
      if (!info) {
        lock[hostname] = null
        return load(domains, opts)
      }
      return info
    })

  lock[hostname] = hostLock
  return hostLock

  function defaults (key) {
    return LE[key]
      .replace(':config', configDir)
      .replace(':hostname', hostname)
  }
}


lib/readCert.js
"use strict";

var fs          = require("mz/fs");
var cp          = require("mz/child_process");
var path        = require("path");
var register    = require("./register");
var CONSTANTS   = require("./constants");
var CERT_PATH   = CONSTANTS.CERT_PATH;
var EXPIRE_TIME = CONSTANTS.EXPIRE_TIME;
var lock        = {};
module.exports  = readCert;

/**
 * Attempts to load or create an ssl certificate with LE or self signed.
 *
 * @param {String} hostname
 * @param {Object} opts
 * @return {Promise}
 */
function readCert (hostname, opts) {
	// Lock to ensure we are not fetching a host multiple times.
	if (lock[hostname]) return lock[hostname];

	var certPath = path.join(CERT_PATH, hostname);
	var stdio    = opts.debug ? "pipe" : "ignore";

	return lock[hostname] = fs.stat(certPath)
		.then(function (stats) {
			// Check for expired cert.
			var age = +new Date - stats.mtime;
			if (age > EXPIRE_TIME) throw new Error("AutoSNI: Expired certificate.");

			// Load cert and key.
			return Promise.all([
				fs.readFile(path.join(certPath, "/cert.pem")),
				fs.readFile(path.join(certPath, "/key.pem")),
			]).then(function (data) {
				// Remove cert later.
				setTimeout(function () { delete lock[hostname]; }, EXPIRE_TIME - age);
				return {
					cert:    data[0],
					key:     data[1]
				}
			})
		})
		.catch(function (err) {
			// Attempt to create a new cert.
			return cp.exec(
				"rm -rf " + certPath + " && mkdir " + certPath,
				{ stdio: stdio }
			)
				.then(function () {
					// Create self signed cert and private key.
					return cp.exec(
						"openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 90 -nodes -subj '/CN=" + hostname + "'",
						{ cwd: certPath, stdio: stdio }
					)
						// Attempt to create LE cert.
						.then(function () { return register(hostname, opts) })
						.catch(function (err) {
							console.error("AutoSNI Error:", err);
							console.error("Will fall back to self signed certificate.");
						});
				})
				.then(function () {
					// After a cert has been made we try to read it again.
					delete lock[hostname];
					return readCert(hostname, opts);
				});
		});
}



lib/register.js
"use strict";

var fs             = require("fs");
var path           = require("path");
var LE             = require("letsencrypt");
var CONSTANTS      = require("./constants");
var CERT_PATH      = CONSTANTS.CERT_PATH;
var STATIC_PATH    = CONSTANTS.STATIC_PATH;
var CHALLENGE_PATH = CONSTANTS.CHALLENGE_PATH;
module.exports     = register;

// Handles for ACME challenges.
var handlers = {
	setChallenge: function (args, key, value, cb) {
		var keyfile = path.join(CHALLENGE_PATH, key);
		fs.writeFile(keyfile, value, "utf8", cb);
	},
	removeChallenge: function (args, key, cb) {
		var keyfile = path.join(CHALLENGE_PATH, key);
		fs.unlink(keyfile, cb);
	}
};

/**
 * Attempts to register a hostname with letsencrypt.
 *
 * @param {String} hostname
 * @param {Object} opts
 * @return {Promise}
 */
function register (hostname, opts) {
	var dir = path.join(CERT_PATH, hostname);
	var port = Number(process.env.PORT || 80);
	return new Promise(function (accept, reject) {
		return LE.create({
			manual:        true,
			debug:         opts.debug,
			configDir:     opts.configDir,
			privkeyPath:   path.join(dir, "key.pem"),
			chainPath:     path.join(dir, "cert.pem"),
			fullchainPath: path.join(dir, "cert.pem"),
			certPath:      path.join(dir, "cert.pem")
		}, handlers).register({
			agreeTos:     true,
			domains:      [hostname],
			email:        opts.email,
			duplicate:    opts.duplicate,
			webrootPath:  STATIC_PATH,
			http01Port:   port,
			tlsSni01Port: port,
			server:       process.env.NODE_ENV === "production"
				? "https://acme-v01.api.letsencrypt.org/directory"
				: "https://acme-staging.api.letsencrypt.org/directory"
		}, function (err, results) {
			if (err) return reject(err);
			accept(results);
		})
	});
}


lib/index.js
'use strict'

var path = require('path')
var LE = require('greenlock-express')
var DEFAULT_DIR = path.join(require('os').homedir(), '/letsencrypt/etc')
var concat = Array.prototype.concat
var noop = function () {}

// Expose lib to the world.
module.exports = createServer

/**
 * Like nodes https.createServer but will automatically generate certificates from letsencrypt
 * falling back to self signed.
 *
 * @param {Object} opts
 * @param {Function} handler
 * @return {Server}
 */
function createServer (opts, app) {
  opts = opts || {}
  app = app || noop
  var handler = null
  var approveDomains = opts.domains
  var ports = opts.ports = opts.ports || {}
  if (!('http' in ports)) ports.http = 80
  if (!('https' in ports)) ports.https = 443
  if (!opts.email) throw new TypeError('AutoSNI: Email is required.')
  if (!opts.agreeTos) throw new TypeError('AutoSNI: Must agree to LE TOS.')
  if (Array.isArray(approveDomains)) {
    // Flatten nested domains.
    approveDomains = concat.apply([], opts.domains || [])
    if (!approveDomains.length) {
      throw new TypeError('AutoSNI: You must specify at least one domain.')
    }
  } else if (typeof approveDomains !== 'function') {
    throw new TypeError('AutoSNI: Domains option must be an array or a function.')
  }

  if (typeof app === 'function') {
    // Allow passing in handler function directly.
    handler = app
  } else if (typeof app.emit === 'function') {
    // Allow passing in another node server (forwards requests).
    handler = function (req, res) {
      app.emit('request', req, res)
    }
  } else {
    throw new TypeError('AutoSNI: Invalid app provided.')
  }

  return LE.create({
    app: handler,
    debug: opts.debug,
    email: opts.email,
    agreeTos: opts.agreeTos,
    approveDomains: approveDomains,
    configDir: opts.dir || DEFAULT_DIR,
    server: opts.debug ? 'staging' : 'production'
  }).listen(opts.ports.http, opts.ports.https)
}
